library/core/src/mem/mod.rs

```diff
@@ -874,6 +874,38 @@ pub const fn replace<T>(dest: &mut T, src: T) -> T {
     }
 }
 
+/// Replaces the value in `dest` by the value returned by `f`.
+///
+/// `f` receives the current referenced value of `dest` as its argument.
+///
+/// * If you want to replace the values of two variables, see [`swap`].
+/// * If you want to replace with a default value, see [`take`].
+/// * If you want replace without the function, see [`replace`].
+///
+/// # Examples
+///
+/// A simple example:
+///
+/// ```
+/// use std::mem;
+///
+/// let mut v: Vec<i32> = vec![1, 2];
+///
+/// mem::replace(&mut v, |mut v| { v.push(3); v });
+/// assert_eq!(vec![1, 2, 3], v);
+/// ```
+#[inline]
+#[unstable(feature = "mem_reshape", issue = "130848")]
+pub fn reshape<T>(dest: &mut T, f: impl FnOnce(T) -> T) {
+    // SAFETY: We read from `dest` but directly write to it afterwards,
+    // such that the old value is not duplicated. Nothing is dropped and
+    // nothing here can panic.
+    unsafe {
+        let result = ptr::read(dest);
+        ptr::write(dest, f(result));
```

Comment:
this is not sound if `f` unwinds, as the value will be dropped twice. see https://docs.rs/replace_with/latest/replace_with/ for details. the design space here is nontrivial, as there are many possible behaviors to do on unwind. this should at least go through an [ACP](https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html), but probably better for more general discussion on internals.rust-lang.org. i'd also suggesting searching there for previous discussion about "replace_with", which is what this function is generally called

Reply:
yeah I just detected that myself and ran into a wall. Sounds good :+1:

Reply:
has there been any discussion about a feature to ensure a closure does not panic?

Reply:
hm. this has probably come up in vague ways in many places but nothing concrete.

Reply:
either way it's definitely both that will arrive any time soon, so it probably makes sense to close the PR and tracking issue for now

Reply:
thank you for it anyways :3

Reply:
:+1:

Reply:
Could you execute the closure inside an extern "C" wrapper func that would abort on unwind? I think I recently saw a PR on a helper that exposed abort on unwind (mirroring the catch_unwind function) using that technique

Reply:
You're thinking of https://github.com/rust-lang/rust/pull/130339